assembly_uploader/ena_queries.py

```diff
@@ -86,36 +54,136 @@ def parse_accession(accession):
+        self.private_url = "https://www.ebi.ac.uk/ena/submit/report/"
+        self.public_url = "https://www.ebi.ac.uk/ena/portal/api/search"
         self.accession = accession
         self.acc_type = parse_accession(accession)
-        username = username or os.getenv("ENA_WEBIN")
-        password = password or os.getenv("ENA_WEBIN_PASSWORD")
+        username = os.getenv("ENA_WEBIN")
+        password = os.getenv("ENA_WEBIN_PASSWORD")
+        if username is None or password is None:
+            logging.error("ENA_WEBIN and ENA_WEBIN_PASSWORD are not set")
         if username and password:
             self.auth = (username, password)
         else:
             self.auth = None
-        self.data = build_data(self.accession, self.acc_type)
-
-    def post_request(self):
-        if self.auth:
-            response = requests.post(
-                self.url,
-                data=self.data,
-                auth=self.auth,
-                **get_default_connection_headers(),
-            )
-        else:
-            logging.warning(
-                "Not authenticated, fetching public data... check ENA_WEBIN and ENA_WEBIN_PASSWORD are set in your "
-                "environment to access private data."
-            )
-            response = requests.post(
-                self.url, data=self.data, **get_default_connection_headers()
-            )
+        self.private = private
+
+    def post_request(self, data):
+        response = requests.post(
+            self.public_url, data=data, **get_default_connection_headers()
+        )
         return response
 
+    def get_request(self, url):
+        response = requests.get(url, auth=self.auth)
+        return response
+
+    def get_data_or_handle_error(self, response):
+        try:
+            data = json.loads(response.text)[0]
+            if data is None:
+                if self.private:
+                    logging.error(
+                        f"{self.accession} private data is not present in the specified Webin account"
+                    )
+                else:
+                    logging.error(f"{self.accession} public data does not exist")
+            else:
+                return data
+        except NoDataException:
+            logging.error(f"Could not find {self.accession} in ENA")
+        except (IndexError, TypeError, ValueError, KeyError):
+            logging.error(
+                f"Failed to fetch {self.accession}, returned error: {response.text}"
+            )
+
+    def retry_or_handle_request_error(self, request, *args, **kwargs):
+        attempt = 0
+        while attempt < RETRY_COUNT:
+            try:
+                response = request(*args, **kwargs)
+                response.raise_for_status()
+                return response
+            #   all other RequestExceptions are raised below
+            except (Timeout, ConnectionError) as retry_err:
+                attempt += 1
+                if attempt > RETRY_COUNT:
```

Comment:
```suggestion
                if attempt >= RETRY_COUNT:
```
I think on third attempt, `attempt==2`, so after `attempt += 1`, `attempt==RETRY_COUNT==3` and we should raise error then?